Lantiq: xrx200: Add support for AVM Fritzbox 3390

cocktailyogi opened the pull request:
All the work has been done some time ago by Andreas Böhler alias @andyboeh  and was already documented in this closed Pull-Request: https://github.com/openwrt/openwrt/pull/2662

I have rebased that changes to Kernel 5.x, tested it and are now filing this pull request, because andyboeh did not want to continue this PR. I added some HW-details here:
https://openwrt.org/inbox/toh/avm/avm_fritzbox_3390

Status:
avm_fritz3390 is working with 5 GHz WIFI and tested for a couple of weeks as DSL-Router

2.4GHz not supported
(2.4 Ghz is implemented by a second SOC via "WASP-Interface". andyboeh did basic work on it, but to me it seemed to complicated and I am not interested in 2.4 Ghz WIFI of this box.) For details see https://www.aboehler.at/doku/doku.php/projects:fritz3390

andyboeh committed:
lantiq: add basic support for AVM FRITZ!Box 3390

The FRITZ!Box 3390 actually contains two SoCs, one Lantiq with a
5GHz WiFi and one AR9342 with a 2.4GHz WiFi. Only the Lantiq
has access to the flash memory, the Atheros runs fully from RAM.

Specifications
--------------

  - Lantiq 500 MHz
  - 128MiB RAM
  - 128MiB NAND
  - 256k Flash
  - AR9580 5GHz WiFi
  - AR9342 560 MHz
  - 64MiB RAM
  - AR9328 2.4GHz WiFi

Remarks
-------

This commit only adds support for the Lantiq side of things and
prepares the drivers for communication with the Atheros SoC. Thus,
only 5GHz WiFi works by default, the 2.4GHz WiFi will be added via
another target.

Some kernel patches will be required to add support for the Atheros SoC.

Installation
------------

Use the eva_ramboot.py script to boot the initramfs image. Then, transfer
the sysupgrade image to the device and run sysupgrade to flash it to the
NAND.

Signed-off-by: Andreas Böhler <dev@aboehler.at>

-----------------------------------------------------------

Rebased support for AVM Fritzbox 3390
Tested as DSL-Router with 5 GHz Wifi.
2.4GHz Wifi is not functiional. 2.4Ghz is implemented in HW via 2nd SOC, which is booted via internal Ethernet,
called WASP-Interface, which is not covered by this Commits

Changed wpad-basic-wolfssl as default

add "pinctrl-0" nand settings. Now it is similar to Fritzbox 3370 with Micron-NAND, which has similar architecture and Nand

remove gpio-setting for WASP-reset

removed unused nand-device

changed ethernet mode to rgmii

Corrected NAND Bank width according to specs

Signed-off-by: Joachim Cerny <cocktail_yogi@web.de>

abajk commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,326 @@
+/dts-v1/;
```
Add Missing SPDX License identifier

abajk replied:
Drop this line:
```
/dts-v1/;
```
Ref: 7f8a1fc55ef77f7d916114dfbdc9460192600fd1

abajk commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,326 @@
+		compatible = "gpio-leds";
+
+		led_power_green: power_green {
+			label = "fritz3390:green:power";
```
Drop model name from LED label
Ref: c3319322b2eec0769a380a14094c3c7482387a03

abajk commented on target/linux/lantiq/xrx200/base-files/etc/board.d/01_leds:
```diff
@@ -48,6 +48,9 @@ avm,fritz3370-rev2-hynix|\
 avm,fritz3370-rev2-micron)
 	ucidef_set_led_switch "lan" "LAN" "green:lan" "switch0" "0x17"
 	;;
+avm,fritz3390)
```
Merge it with previous case

abajk commented on target/linux/lantiq/xrx200/base-files/etc/board.d/02_network:
```diff
@@ -47,6 +47,9 @@ lantiq_setup_interfaces()
 			"2:lan:3" "3:lan:4" "4:lan:1" "5:lan:2" "6t@eth0"
+	avm,fritz3390)
+		ucidef_add_switch "switch0" \
+			"0:lan:3" "1:lan:4" "2:lan:2" "4:lan:1" "5:lan:5" "6t@eth0"
```
Is port number 5 connected to WiFi?

cocktailyogi replied:
Yes, from what I understood it seems to be the connection to the second SOC (also know as WASP-interface)

andyboeh replied:
I would remove LAN5 entirely for now, as it causes more confusion than it serves - see my other comments.

abajk commented on target/linux/lantiq/xrx200/base-files/etc/board.d/02_network:
```diff
@@ -135,7 +139,8 @@ lantiq_setup_macs()
 	avm,fritz7360sl)
 		wan_mac=$(macaddr_add "$(mtd_get_mac_binary urlader 0xa91)" 1)
-		;;
+		;;	
```
Remove trailing whitespace

abajk commented:
Maybe squash some commits

cocktailyogi commented:
Okay, I have integrated your suggestions.
But I was not able to squash my commits. I googled how to squash commits on Github, but I am not sure if I understand how to do it. I am afriad that I might bring in this more mess, than it helps. So could someone guide me?

Yogi

jschwartzenberg commented:
> But I was not able to squash my commits. I googled how to squash commits on Github, but I am not sure if I understand how to do it. I am afriad that I might bring in this more mess, than it helps.

You have to do this using Git (not GitHub). First rebase using:
`git rebase -i <parent commit>`

Then in your text editor you can use squash/fixup.

Then:
`git push -f`

You have to use `-f` for a forced push as you're rewriting history. (The alternative would be a new branch & new PR, but GitHub handles squashes & forced pushes to existing branches well in PRs.)

In case you want to keep the original commit set around as a precaution, create another branch first and then switch back to this one before you squash & push.

abajk commented on target/linux/lantiq/image/vr9.mk:
```diff
@@ -118,6 +118,18 @@ define Device/avm_fritz3370-rev2-micron
+  SOC := vr9
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 49152k
+  DEVICE_PACKAGES := kmod-ath9k kmod-owl-loader wpad-basic kmod-usb-dwc2 \
```
Replace wpad-basic with wpad-basic-wolfssl (a14f5bb4bd263c21e103f13279d0c2ff03e48fe5).

abajk commented on target/linux/lantiq/image/vr9.mk:
```diff
@@ -118,6 +118,18 @@ define Device/avm_fritz3370-rev2-micron
+  $(Device/AVM)
+  $(Device/NAND)
+  DEVICE_MODEL := FRITZ!Box 3390
+  SOC := vr9
```
SOC:=9 is unnececary. Default SoC is already set in openwrt/target/linux/lantiq/image/Makefile.

cocktailyogi commented:
> 
> 
> > But I was not able to squash my commits. I googled how to squash commits on Github, but I am not sure if I understand how to do it. I am afriad that I might bring in this more mess, than it helps.
> 
> You have to do this using Git (not GitHub). First rebase using:
> `git rebase -i <parent commit>`
> 
> Then in your text editor you can use squash/fixup.
> 
> Then:
> `git push -f`
> 
> You have to use `-f` for a forced push as you're rewriting history. (The alternative would be a new branch & new PR, but GitHub handles squashes & forced pushes to existing branches well in PRs.)
> 
> In case you want to keep the original commit set around as a precaution, create another branch first and then switch back to this one before you squash & push.

Okay,

I did what you suggested. Now all the changes including recommendations from @abajk are in the last commit
https://github.com/openwrt/openwrt/pull/3656/commits/ef5836b3a3e0780b078565e558efa9f4c7e28fe0

But all the other little commits I have squashed still show up here in this PR. I am not sure if this is right.

Yogi

cocktailyogi commented:
> 
> 
> Maybe squash some commits

Done :-)

Thx to @jschwartzenberg

adschm commented:
This should be tidied up:
- Create commit by what they do, not by their history (e.g. one commit for fixing arch-wide gpio, one commit for adding devices, one commit for another separate subject)
- Add missing commit messages
- Take care to properly show/document authorship
- There is no kernel 4.19 support anymore

Apart from that, be aware that the last PR was closed due to lack of interest, so it's probably that this one will meet the same fate and your work might be in vain (apart from individuals using the PR as is)

cocktailyogi commented:
>     * There is no kernel 4.19 support anymore
I know, but what does this mean here. I did not include anything regarding Kernel 4.19

>     * Take care to properly show/document authorship
Is done, I think

>     * I just had squasehd my commits, because it was recommented. Now ther eis ne commit, what I did:
Rebasing the work somebody else has done. Basically this is one step.

andyboeh commented:
Why are you basing your work on my avm_fritz3390_wifi branch instead of the avm_fritz3390 branch? The _wifi branch has all the infrastructure in place to bring up WiFi, whereas the other branch only includes the stuff you are actually trying to get included. 

The problem here is that your commits first add a lot of stuff that you remove afterwards. So basically, your PR should come up with two commits:

  - One that adds the `gpio-ranges` define for all VR9 devices
  - One that adds device support

See #2662 for reference, where exactly this is done.

andyboeh commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,326 @@
+			gpios = <&gpio 34 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
```
This export is only required if you bring up the secondary WiFi SoC. Personally, I would remove it for now - please wait for other reviewer's comments before doing so.

cocktailyogi replied:
Okay, we should decide about this.

abajk replied:
Also I woud prefer to remove this node.

andyboeh commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,326 @@
+				speed = <1000>;
+				full-duplex;
+			};
+		};
```
This node is only required for the WiFI SoC. On 4.19, this had no effect without a further patch because the xrx200 driver did not support fixed-link nodes. I don't know about 5.4 so please verify that.

cocktailyogi commented:
> 
> 
> Why are you basing your work on my avm_fritz3390_wifi branch instead of the avm_fritz3390 branch? The _wifi branch has all the infrastructure in place to bring up WiFi, whereas the other branch only includes the stuff you are actually trying to get included.
> 
> The problem here is that your commits first add a lot of stuff that you remove afterwards. So basically, your PR should come up with two commits:
> 
>     * One that adds the `gpio-ranges` define for all VR9 devices
> 
>     * One that adds device support
> 
> 
> See #2662 for reference, where exactly this is done.

Hi Andy, thx for your reply.
The answer is easy: Because I am beginner and not used to this big Github-Community. Is just wanted to use my Fritzbox3390 on Openwrt and make your work available for everyone in the future. This is easier, when it is fully comitted into Openwrt. I had it, if work is done and nobody respects it. Sorry that is used your wrong branch. 
Do you think, I should close this PR and redo the job based on your other branch and file an new clean PR? I can do this, if this helps. 

Yogi

andyboeh commented:
> The answer is easy: Because I am beginner and not used to this big Github-Community. Is just wanted to use my Fritzbox3390 on Openwrt and make your work available for everyone in the future. This is easier, when it is fully comitted into Openwrt. I had it, if work is done and nobody respects it. Sorry that is used your wrong branch.

No worries, it just would have saved you a lot of work.

> Do you think, I should close this PR and redo the job based on your other branch and file an new clean PR? I can do this, if this helps.

No, just squash it properly - same effect.

cocktailyogi commented:
Okay, I squashed it more. Now it looks like the Wifi-Stuff was newer in here.

adschm commented:
> I know, but what does this mean here. I did not include anything regarding Kernel 4.19

target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts ?

adschm commented:
> Okay, I squashed it more. Now it looks like the Wifi-Stuff was never in here.

As both andyboeh and me stated: This needs only two commits: one for the gpio-ranges, and one for device support.

Keeping somebody elses work does not mean you cannot change it anymore, you just need to document what you changed in the commit message ...

cocktailyogi commented:
> 
> 
> > I know, but what does this mean here. I did not include anything regarding Kernel 4.19
> 
> target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts ?

Sorry, where do you see that file? in my branch there is no folder "files-4.19" any more.

cocktailyogi commented:
> 
> 
> > Okay, I squashed it more. Now it looks like the Wifi-Stuff was never in here.
> 
> As both andyboeh and me stated: This needs only two commits: one for the gpio-ranges, and one for device support.
> 
> Keeping somebody elses work does not mean you cannot change it anymore, you just need to document what you changed in the commit message ...

Better now?

adschm commented:
Yes, that's the idea. Now we can start talking about details.

BTW, first commit still lacks a commit message.

jschwartzenberg commented:
> Apart from that, be aware that the last PR was closed due to lack of interest

Could you highlight what kind of interest would be necessary? Unfortunately I do not have this particular model, but my ISP provides a related model and I am aware quite a few people would be interested in mainlined OpenWRT support.

adschm commented:
> > Apart from that, be aware that the last PR was closed due to lack of interest
> 
> Could you highlight what kind of interest would be necessary? Unfortunately I do not have this particular model, but my ISP provides a related model and I am aware quite a few people would be interested in mainlined OpenWRT support.

There was just nobody interested in reviewing it.

andyboeh commented:
> > > Apart from that, be aware that the last PR was closed due to lack of interest
> > 
> > 
> > Could you highlight what kind of interest would be necessary? Unfortunately I do not have this particular model, but my ISP provides a related model and I am aware quite a few people would be interested in mainlined OpenWRT support.
> 
> There was just nobody interested in reviewing it.

And finally, I was no longer interested in rebasing and porting things forward. In the meantime, I have no use for routers with VDSL modem any more.

cocktailyogi commented:
In my opinion AVM Fritzbox 3390 is nice because it has manageable Gbit-Switch, 4 ports, average performance and RAM. In addition 5Ghz and 2.4GHz Wifi. Even without using the internal DSL-modem, converted to Openwrt this is a nice Router with optional features. And because the device is older, some people like me might have an unused one. Then it is a nice option without spending money. 
In addition this Fritzbox was over the years running much more stable than other AVM products I have seen.

cocktailyogi commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9.dtsi:
```diff
@@ -223,6 +223,7 @@
 			compatible = "lantiq,xrx200-pinctrl";
 			#gpio-cells = <2>;
 			gpio-controller;
+			gpio-ranges = <&gpio 0 0 56>;
```
You might be right. I have tested it without the gpio-ranges line definition and it is getting gpio-errors during boot.
I was not able to find any gpio number greater than 46 so there should be not problem setting gpio-ranges to 50. A test was successful.

mrkiko commented:
Please keep up with the good work man!! And thank you a lot for it!
I didn't manage to get Ethernet working on my FRITZ!BOX 3272.... So was looking this PR to see if there where some similarities I could pick up from. :)

thanks!

abajk commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,326 @@
+		compatible = "lantiq,nand-xway";
+		bank-width = <2>;
+		reg = <1 0x0 0x2000000>;
+
```
~~~
<&nand_cs1_pins>
~~~
 It is only needed when
~~~
lantiq,cs = <1>;
~~~
So you can remove it.

cocktailyogi committed:
lantiq: VR9: fix gpio-hog by defining the GPIO ranges

defined gpio-ranges according to Reference: https://github.com/torvalds/linux/blob/master/drivers/pinctrl/pinctrl-xway.c#L864
This change is needed by gpio-hog on Fritzbox 3390
thx to @abajk

Signed-off-by: Joachim Cerny <cocktail_yogi@web.de>

abajk commented:
> 
> 
> Please keep up with the good work man!! And thank you a lot for it!
> I didn't manage to get Ethernet working on my FRITZ!BOX 3272.... So was looking this PR to see if there where some similarities I could pick up from. :)
> 
> thanks!
@mrkiko
Is the FRITZ!BOX 3272 based on AR10?
This SoC use same switch version as grx390(GSWIP 2.2).
Look at my PR for DWR966[1] and mailnie effort [2][3][4]. During my tests, Ethernet USB and the first PCIe worked. Switch only needs a few minor changes.

[1] https://github.com/openwrt/openwrt/pull/3102
[2] https://github.com/torvalds/linux/commit/58c9e24721c4a84eb5a6db3c1d54dba97e97b3f7#diff-a06e2d399118d4a48f8e4ce880871b459e39f2b3d910861160826135a2e65570
[3] https://github.com/torvalds/linux/commit/329afb94e0052fd7a94652d1a14df0b8f5527413#diff-c833408ad2dfc733759f33f31aa23fa0599282ad5ba47e237a8c074efd59f354
[4] https://lkml.org/lkml/2020/12/6/304

cocktailyogi commented:
Okay, I have included all your requested changes. Also build and tested it. Runs good :-)

So ready to race from my side.

abajk commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,308 @@
+		ethernet@0 {
+			compatible = "lantiq,xrx200-pdi-port";
+			reg = <0>;
+			phy-mode = "sgmii";
```
The VR9 support only MII, RMII or RGMII on port 0 and 1.
Change this to rgmii.

abajk replied:
DTS generally looks good for me.

abajk commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,308 @@
+		ethernet@1 {
+			compatible = "lantiq,xrx200-pdi-port";
+			reg = <1>;
+			phy-mode = "sgmii";
```
rgmii

cocktailyogi replied:
done

abajk replied:
The mode should be changed to rgmii.

cocktailyogi commented:
Okay, I think all requested changes are done. :-)

adschm commented:
You should condense the list of changes in the first commit message. One Signed-off from you is enough, and all changes you did should be listed before that (after the other person's Signed-off).

adschm commented on target/linux/lantiq/image/vr9.mk:
```diff
@@ -118,6 +118,16 @@ define Device/avm_fritz3370-rev2-micron
+  DEVICE_MODEL := FRITZ!Box 3390
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 49152k
+  DEVICE_PACKAGES := kmod-ath9k kmod-owl-loader wpad-basic-wolfssl kmod-usb-dwc2 fritz-tffs
```
Please wrap lines with more than 74 characters

cocktailyogi commented:
> 
> 
> You should condense the list of changes in the first commit message. One Signed-off from you is enough, and all changes you did should be listed before that (after the other person's Signed-off).

All done :-)

cocktailyogi commented:
Okay, rgmii-thing on eth1 is fixed. Sorry, I have overlooked that setting.

cocktailyogi commented:
Okay, whats next? From what I understand I need 2 reviews. From what I understand @abajk und @adschm did an intensive review.

Kind regards,
Yogi

adschm commented:
One of the commits still doesn't have a commit message.

Apart from that, you need somebody doing a full review and/or a committer interested in merging this (and invest further time).

cocktailyogi commented:
> 
> 
> One of the commits still doesn't have a commit message.
> 
> Apart from that, you need somebody doing a full review and/or a committer interested in merging this (and invest further time).

Okay, sounds good. 
When we have found reviewers/commiters who are interested in merging this, I will inverst more time and fix that missing comment. ;-)
Have a good and safe Christmas-time.

Yogi

adschm commented:
> When we have found reviewers/commiters who are interested in merging this, I will invest more time and fix that missing comment. ;-)

Well, it's typically the other way around: If you want to get a reviewer where they are scarce, you should polish everything so it's easy for them to review (which will also save you a few additional changes-requested/review cycles if you find one).

jschwartzenberg commented:
> One of the commits still doesn't have a commit message.

I see commit messages for both commits. What is expected?


> When we have found reviewers/commiters who are interested in merging this, I will invest more time and fix that missing comment. ;-)

Yeah as mentioned it's really the best to polish this as much as possible as soon as possible. A fully polished PR should attract reviews/approvals. I am also curious however, @adschm, I see two approvers are required. Would you be comfortable to be one of them (with the commit message completed)?

cocktailyogi commented:
> 
> 
> One of the commits still doesn't have a commit message.
> 
Which one? On my screen they both have commit-messages.

adschm commented:
> > One of the commits still doesn't have a commit message.
> 
> Which one? On my screen they both have commit-messages.

lantiq: VR9: fix gpio-hog by defining the GPIO ranges

cocktailyogi commented:
> 
> 
> > > One of the commits still doesn't have a commit message.
> > 
> > 
> > Which one? On my screen they both have commit-messages.
> 
> lantiq: VR9: fix gpio-hog by defining the GPIO ranges

added more text ;-)

abajk commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts:
```diff
@@ -0,0 +1,308 @@
+&localbus {
+	flash@1 {
+		compatible = "lantiq,nand-xway";
+		bank-width = <2>;
```
Documantation: 
~~~
 - bank-width : Width (in bytes) of the bank.  Equal to the
   device width times the number of interleaved chips.
~~~
MICRON MT29F1G08ABADAWP is 8bit memory. Try this:
~~~
bank-width = <1>;
~~~

cocktailyogi replied:
Okay, you might be right, but I am not sure if it is worth to change and test it. 
It is same Micron-Flash like in some variants of Fritzbox 3370, which is offically supported by Openwrt. In file vr9_avm_fritz3370-rev2-micron.dts it is set to "bank-width = <2>;"
And my box is running well with this "wrong" setting. If we change it, we should correct it in all dts files, that are using this Micron-Flash.
What do you think?

jschwartzenberg replied:
I would say test it on your box for this model. For other models a separate PR (or even separate PRs) had better be created and separately tested.

cocktailyogi replied:
Done

cocktailyogi replied:
done

abajk commented:
> > > > One of the commits still doesn't have a commit message.
> > > 
> > > 
> > > Which one? On my screen they both have commit-messages.
> > 
> > 
> > lantiq: VR9: fix gpio-hog by defining the GPIO ranges
> 
> added more text ;-)

It would be good to explain in the commit description that this change is needed by gpio-hog on Fritzbox 3390.

cocktailyogi commented:
> 
> 
> > > > > One of the commits still doesn't have a commit message.
> > > > 
> > > > 
> > > > Which one? On my screen they both have commit-messages.
> > > 
> > > 
> > > lantiq: VR9: fix gpio-hog by defining the GPIO ranges
> > 
> > 
> > added more text ;-)
> 
> It would be good to explain in the commit description that this change is needed by gpio-hog on Fritzbox 3390.

done

abajk commented:
Acked-by: Aleksander Jan Bajkowski <A.Bajkowski@stud.elka.pw.edu.pl>

cocktailyogi commented:
Okay, whats next? I have done all requested changes. Now its time to finish things.

jschwartzenberg commented:
> Okay, whats next? I have done all requested changes. Now its time to finish things.

At least one more code review is necessary, but to obtain the best quality it never hurts to have even more people look at the changes. Maybe you could ask in #openwrt on Freenode, ask through the mailing list or the forum?

Hopefully a second reviewer can at least join in!

cocktailyogi commented:
@adschm 
Could you please have a look at the final chnages? You have done a lot of review-work and helped a lot. It would be nice, if you could approve this.

Yogi

cocktailyogi commented:
> 
> 
> > Okay, whats next? I have done all requested changes. Now its time to finish things.
> 
> At least one more code review is necessary, but to obtain the best quality it never hurts to have even more people look at the changes. Maybe you could ask in #openwrt on Freenode, ask through the mailing list or the forum?
> 
> Hopefully a second reviewer can at least join in!

Why cannot you do the codereview?

jschwartzenberg commented:
> Why cannot you do the codereview?

I'm not an OpenWRT developer. I just hope to see this merged because Fritzboxes are also commonly used where I live.

People that are part of the OpenWRT GitHub organization need to review & approve I think.

cocktailyogi commented:
> 
> 
> > Why cannot you do the codereview?
> 
> I'm not an OpenWRT developer. I just hope to see this merged because Fritzboxes are also commonly used where I live.
> 
> People that are part of the OpenWRT GitHub organization need to review & approve I think.

Merry Christmas,

you are a helpful guy and have good knowledge. they should promote you into developer status. In our area people are using plenty of AVM-devices. Lets hack them :-)

jschwartzenberg commented:
> Merry Christmas,

Merry Christmas to you too and/or Merry Grav-Mass ;)
Hope everybody can enjoy these days!


> you are a helpful guy and have good knowledge. they should promote you into developer status. In our area people are using plenty of AVM-devices. Lets hack them :-)

Haha not really. I know Git and while I am a software developer, C is not a language I know well and I know little about OpenWRT as well. Even if they would promote me so that I could approve a PR like this, I would never do it.

We need to be patient here. OpenWRT and also the Linux kernel could only be where they are now because people took care to make sure changes going in were of high quality. Code reviews are an important part of software development. At work I am known for stating that while a feature might already implemented and in the code review phase, that could still mean it is far from done as during the code review it might turn out to be necessary to rewrite the entire implementation of it.

Do not be discouraged! Having one approval for this is already a major milestone. It is not unlikely you will be asked to change some things before there will be a second approval. This can be a long process, but it is only so much for the better once your changes end up in OpenWRT. As all changes go through these processes, it ensures that once this router model is supported it will continue to work in the future as others bring in their changes. Having well-understood code also helps new developers to step in and contribute.

With typical proprietary firmware, the vendors have access to all hardware to be supported so they can actually test each and every release they put out on all their models. OpenWRT does not have this, so code quality is much more important.

So get this further going, I would recommend asking around in the OpenWRT community as I suggested. Being acquainted there will also create trust. Even when changes are merged in, it is possible some after-care is needed too. Showing interest and being involved with the OpenWRT community also creates the necessary trust that you will be around when such after-care is needed. It is not possible to expect OpenWRT project members to reach out to contributors when facing issues. They rely on being able to understand issues themselves, but having original contributors in reach is ideal. Can they expect you to monitor their bug tracker from time to time and reply to Fritzbox 3390 issues for example?

A second thing I recommend doing is contacting your ISP and/or AVM. Explain them the benefits you see in them supplying resources. If AVM as a vendor would at least provide minimal resources to OpenWRT development, it would already be great.

I hope this helps! Good luck!!

abajk commented:
@cocktailyogi 
It looks like gpio-ranges is no more needed since kernel 5.0.

[1] https://github.com/abajk/linux/commit/9b4924da4711674e62d97d4f5360446cc78337af

cocktailyogi commented:
Hi, I have tested it three weeks ago. On my Box Fritzbox 3390 it is needed.
Without GPIO-Ranges, GPIO-numbers are getting messed up.

Am Do., 7. Jan. 2021 um 21:36 Uhr schrieb Aleksander Bajkowski <
notifications@github.com>:

> @cocktailyogi <https://github.com/cocktailyogi>
> It looks like gpio-ranges is no more needed since kernel 5.0.
>
> [1] abajk/linux@9b4924d
> <https://github.com/abajk/linux/commit/9b4924da4711674e62d97d4f5360446cc78337af>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/3656#issuecomment-756369073>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABX5OZWPCIOLMB5JGWJS6V3SYYLONANCNFSM4UPGOUZA>
> .
>

hauke commented:
Thank you for the pull request, I applied it to master.

jschwartzenberg commented:
Great to see progress on the AVM boxes!! Thanks a lot everybody!

Gisi0 commented:
Great to have an image for this, I tried to install the snapshot and get this error using eva_ramboot.py
I hope somebody can help me.

`SETENV memsize 0x07ae4000
Traceback (most recent call last):
  File "./eva_ramboot.py", line 40, in <module>
    adam('SETENV memsize 0x%08x'%(addr))
  File "./eva_ramboot.py", line 33, in adam
    resp = ftp.sendcmd(cmd)
  File "/usr/lib/python3.7/ftplib.py", line 273, in sendcmd
    return self.getresp()
  File "/usr/lib/python3.7/ftplib.py", line 246, in getresp
    raise error_perm(resp)
ftplib.error_perm: 500 'SETENV memsize 0x07ae4000': command not understood.`

sfritz2 commented:
Works fine for me. If it is not working for you, flash a 3370 image (https://openwrt.org/toh/avm/fritz.box.wlan.3370?s[]=3370 ), after make a sysupgrade to 3390.

sfritz2 commented:
also try to set first the ip to 192.168.178.5 and make a ping on boot, ttl should be 250. if not, restart the fritzbox.

Gisi0 commented:
Got it to run, forgot a switch between the box and my pc. Network was to late up.
